### ISSUE TITLE
Left join seems wrong with multiplicities

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#2724]
Set up two collections to look like `left` and `right`, and then do a left join:
```
materialize=> select * from left;
 column1 | column2 
---------+---------
       1 |       2
       1 |       2
       1 |       3
(3 rows)
materialize=> select * from right;
 key | val 
-----+-----
   1 | a
(1 row)
materialize=> select * from "left" left join "right" on "left".column1 = "right".key;
 column1 | column2 | key | val 
---------+---------+-----+-----
       1 |       2 |     | 
       1 |       2 |   1 | a
       1 |       2 |   1 | a
       1 |       3 |   1 | a
(4 rows)
materialize=> 
```
The first row shouldn't be there.

It looks like this might be an issue in `anti_lookup`, which appears to subtract a single copy of keys independent of the multiplicity in `self`.

**Comment by jamii** [comment]
Same behaviour with optimizations disabled.

**Comment by jamii** [comment]
``` diff
--- a/src/expr/relation/mod.rs
+++ b/src/expr/relation/mod.rs
@@ -993,24 +993,33 @@ impl RelationExpr {
         keys_and_values: RelationExpr,
         default: Vec<(Datum, ColumnType)>,
     ) -> RelationExpr {
-        let keys = self;
-        assert_eq!(keys_and_values.arity() - keys.arity(), default.len());
-        keys_and_values
-            .let_in(id_gen, |_id_gen, get_keys_and_values| {
-                Ok(get_keys_and_values
-                    .distinct_by((0..keys.arity()).collect())
-                    .negate()
-                    .union(keys)
-                    // This join is logically equivalent to
-                    // `.map(<default_expr>)`, but using a join allows for
-                    // potential predicate pushdown and elision in the
-                    // optimizer.
-                    .product(RelationExpr::constant(
-                        vec![default.iter().map(|(datum, _)| *datum).collect()],
-                        RelationType::new(default.iter().map(|(_, typ)| typ.clone()).collect()),
-                    )))
+        assert_eq!(keys_and_values.arity() - self.arity(), default.len());
+        self.let_in(id_gen, |id_gen, get_keys| {
+            keys_and_values.let_in(id_gen, |_id_gen, get_keys_and_values| {
+                Ok(RelationExpr::join(
+                    vec![
+                        get_keys_and_values
+                            .distinct_by((0..get_keys.arity()).collect())
+                            .negate()
+                            .union(get_keys.clone().distinct()),
+                        get_keys.clone(),
+                    ],
+                    (0..get_keys.arity())
+                        .map(|i| vec![(0, i), (1, i)])
+                        .collect(),
+                )
+                .project((0..get_keys.arity()).collect())
+                // This join is logically equivalent to
+                // `.map(<default_expr>)`, but using a join allows for
+                // potential predicate pushdown and elision in the
+                // optimizer.
+                .product(RelationExpr::constant(
+                    vec![default.iter().map(|(datum, _)| *datum).collect()],
+                    RelationType::new(default.iter().map(|(_, typ)| typ.clone()).collect()),
+                )))
             })
-            .unwrap()
+        })
+        .unwrap()
     }
```

**Comment by jamii** [comment]
```
jamie=> create database jamie;
CREATE DATABASE
jamie=> create table foo(a int);
CREATE TABLE
jamie=> create table bar(b int);
CREATE TABLE
jamie=> insert into foo values (1), (1);
INSERT 0 2
jamie=> insert into foo values (2), (2);
INSERT 0 2
jamie=> insert into bar values (1);
INSERT 0 1
jamie=> select * from foo left join bar on a=b;
 a | b 
---+---
 2 |  
 2 |  
 1 | 1
 1 | 1
(4 rows)

jamie=> explain plan for select * from foo left join bar on a=b;
                     Optimized Plan                      
---------------------------------------------------------
 %0 =                                                   +
 | Get jamie.public.foo (u1)                            +
 | ArrangeBy (#0)                                       +
                                                        +
 %1 =                                                   +
 | Get jamie.public.bar (u3)                            +
 | ArrangeBy (#0)                                       +
                                                        +
 %2 =                                                   +
 | Join %0 %1 (= #0 #1)                                 +
 | | implementation = DeltaQuery %0 %1.(#0) | %1 %0.(#0)+
 | | demand = (#0)                                      +
 | Filter !(isnull(#0))                                 +
                                                        +
 %3 =                                                   +
 | Get %2                                               +
 | Project (#0, #0)                                     +
                                                        +
 %4 =                                                   +
 | Get %2                                               +
 | Distinct group=(#0)                                  +
 | Negate                                               +
                                                        +
 %5 =                                                   +
 | Get jamie.public.foo (u1)                            +
 | Distinct group=(#0)                                  +
                                                        +
 %6 =                                                   +
 | Union %4 %5                                          +
                                                        +
 %7 =                                                   +
 | Get jamie.public.foo (u1)                            +
 | ArrangeBy (#0)                                       +
                                                        +
 %8 =                                                   +
 | Join %6 %7 (= #0 #1)                                 +
 | | implementation = Differential %6 %7.(#0)           +
 | | demand = (#0)                                      +
 | Map null                                             +
 | Project (#0, #2)                                     +
                                                        +
 %9 =                                                   +
 | Union %3 %8                                          +
 
(1 row)
```

**Comment by jamii** [comment]
Fixed by #2727
